diag_manager/diag_util.F90

```diff
@@ -731,182 +757,360 @@ SUBROUTINE log_diag_field_info(module_name, field_name, axes, long_name, units,&
-                                        !! equal to <TT>Time</TT> or <TT>Time_zero</TT>.
-    CHARACTER(len=*), INTENT(out) :: err_msg !< Return status of <TT>check_bounds_are_exact_dynamic</TT>.
-                                             !! An empty error string indicates the x, y, and z indices are
-                                             !!     equal to the buffer array boundaries.
+      err_msg = ''
+    END IF
+  end associate
+END SUBROUTINE check_out_of_bounds_legacy
+
+ !> @brief Checks if the array indices for <TT>output_fields(out_num)</TT> are outside the
+  !! <TT>output_fields(out_num)%buffer</TT> upper and lower bounds.
+  !! If there is an error then error message will be filled.
+SUBROUTINE fms_check_out_of_bounds_modern_r4(ofb, bounds, output_name, module_name, err_msg)
+  REAL(kind=r4_kind), INTENT (in), DIMENSION(:,:,:,:,:) :: ofb !< The output field buffer to check
+  TYPE (fms_diag_ibounds_type), INTENT(inout) :: bounds !< The bounding box to check against
+  CHARACTER(:), ALLOCATABLE, INTENT(in) :: output_name !< output name for placing in error message
+  CHARACTER(:), ALLOCATABLE, INTENT(in) :: module_name !< module name for placing in error message
+  CHARACTER(len=*), INTENT(inout) :: err_msg !< Return status of <TT>check_out_of_bounds</TT>.  An empty
+                                           !! error string indicates the x, y, and z indices are not outside the
+
+  CHARACTER(len=128) :: error_string1, error_string2
+  LOGICAL :: out_of_bounds = .true.
+  TYPE (fms_diag_ibounds_type) :: array_bounds
+
+  CALL bounds_from_array(array_bounds, ofb)
+
+  out_of_bounds = compare_buffer_bounds_to_size(array_bounds, bounds, &
+     & error_string2, a_lessthan_b, a_greaterthan_b)
+
+  IF (out_of_bounds .EQV. .true.) THEN
+     WRITE(error_string1,'(a,"/",a)')  TRIM(module_name), TRIM(output_name)
+     err_msg = 'module/output_field='//TRIM(error_string1)//&
+          & '  Bounds of buffer exceeded.  '//TRIM(error_string2)
+     !   imax, imin, etc need to be reset in case the program is not terminated.
+     call bounds%reset(VERY_LARGE_AXIS_LENGTH,0)
+  ELSE
+     err_msg = ''
+  END IF
+END SUBROUTINE fms_check_out_of_bounds_modern_r4
+
+ !> @brief Checks if the array indices for output_field buffer (ofb) are outside the
+  !! are outside the bounding box (bounds).
+  !! If there is an error then error message will be filled.
+
+SUBROUTINE fms_check_out_of_bounds_modern_r8(ofb, bounds, output_name, module_name, err_msg)
+  REAL(kind=r8_kind), INTENT (in), DIMENSION(:,:,:,:,:) :: ofb !< The output field buffer to check
+  TYPE (fms_diag_ibounds_type), INTENT(inout) :: bounds !< The bounding box to check against
+  CHARACTER(:), ALLOCATABLE, INTENT(in) :: output_name  !< output name for placing in error message
+  CHARACTER(:), ALLOCATABLE, INTENT(in) :: module_name   !< module name for placing in error message
+  CHARACTER(len=*), INTENT(out) :: err_msg !< Return status of <TT>check_out_of_bounds</TT>.  An empty
```

Comment:
Its always present and I am trying to follow the existing interface (signature and documentation) of the corresponding routines, which did not have it as optional  and also checked it when the procedure returned since its also  a "return status". But the type intent should be (inout).  In the legacy code, the trio check_out_of_bounds,  check_bounds_are_exact_dynamic check_bounds_are_exact_static mixed the type intent; I'm using inout in all three.